src/material_library.cpp

```diff
@@ -0,0 +1,342 @@
+  pyne::matname_set mats_to_add = mat_lib.get_matlist();
+  for (auto it = mats_to_add.begin(); it != mats_to_add.end(); it++) {
+    pyne::Material mat = Material(mat_lib.get_material(*it));
+    (*this).add_material(mat);
+  }
+}
+
+void pyne::MaterialLibrary::load_json(Json::Value json) {
+  
+  Json::Value::Members keys = json.getMemberNames();
+  Json::Value::Members::const_iterator ikey = keys.begin();
+  Json::Value::Members::const_iterator ikey_end = keys.end();
+  for (; ikey != ikey_end; ++ikey) {
+    pyne::Material mat = pyne::Material();
+    mat.load_json(json[*ikey]);
+    (*this).add_material(Material(mat));
+  }
+}
+
+Json::Value pyne::MaterialLibrary::dump_json() {
+  Json::Value json = Json::Value(Json::objectValue);
+
+  for (auto name : name_order) {
+    json[name] = material_library[name]->dump_json();
+  }
+  return json;
+}
+
+void pyne::MaterialLibrary::from_json(char* fname) {
+  std::string filename(fname);
+  from_json(filename);
+
+}
+void pyne::MaterialLibrary::from_json(const std::string& filename) {
+  if (!pyne::file_exists(filename))
+    throw pyne::FileNotFound(filename);
+  std::string s;
+  std::ifstream f (filename.c_str(), std::ios::in | std::ios::binary);
+  f.seekg(0, std::ios::end);
+  s.resize(f.tellg());
+  f.seekg(0, std::ios::beg);
+  f.read(&s[0], s.size());
+  f.close();
+  Json::Reader reader;
+  Json::Value json;
+  reader.parse(s, json);
+  load_json(json);
+}
+
+void pyne::MaterialLibrary::write_json(char * filename) {
+  std::string fname (filename);
+  write_json(fname);
+}
+
+void pyne::MaterialLibrary::write_json(const std::string& filename) {
+  Json::Value json = dump_json();
+  Json::StyledWriter writer;
+  std::string s = writer.write(json);
+  std::ofstream f;
+  f.open(filename.c_str(), std::ios_base::trunc);
+  f << s << "\n";
+  f.close();
+}
+
+void pyne::MaterialLibrary::add_material(pyne::Material mat) {
+  // if exists, get the material name from metadata make one instead
+  std::string mat_name;
+  std::pair<pyne::matname_set::iterator, bool> mat_insert;
+  if (mat.metadata.isMember("name")) {
+    mat_name = mat.metadata["name"].asString();
+    mat_insert = matlist.insert(mat.metadata["name"].asString());
+  } else {
+    // form a mat name as 'mX'
+    int mat_number = name_order.size() +1;
+    mat_name = "m" + std::to_string(mat_number);
+    mat.metadata["name"] = mat_name;
+    mat_insert = matlist.insert(mat.metadata["name"].asString());
+
+    // if 'mX' name is already in the Library rename the material
+    while (!mat_insert.second) {
+      mat_number += name_order.size();
+      mat_name = "m" + std::to_string(mat_number);
+      mat.metadata["name"] = mat_name;
+      mat_insert = matlist.insert(mat.metadata["name"].asString());
+    }
+  }
+  if (mat_insert.second) {
+    append_to_nuclist(mat);
+    material_library[mat_name] = new Material(mat);
+    material_library[mat_name]->metadata["mat_number"] = int(name_order.size() +1);
```

Comment:
I'm not sure it's wise to change the `mat_number` in the meta data.... but this could be complicated...